manager/state/raft/raft.go

```diff
@@ -524,13 +526,33 @@ func (n *Node) Join(ctx context.Context, req *api.JoinRequest) (*api.JoinRespons
 		}
 	}
 
+	remoteAddr := req.Addr
+
+	// If the joining node sent an address like 0.0.0.0:4242, automatically
+	// determine its actual address based on the GRPC connection. This
+	// avoids the need for a prospective member to know its own address.
+
+	requestHost, requestPort, err := net.SplitHostPort(remoteAddr)
+	if err != nil {
+		return nil, fmt.Errorf("invalid address %s in raft join request", remoteAddr)
+	}
+
+	requestIP := net.ParseIP(requestHost)
```

Comment:
It's fine if `net.ParseIP` fails. That just means we try to "ping" the literal address we were given, instead of potentially substituting the observed address. If someone gives us a bad hostname that we can't resolve, it will fail at the ping step.